src/backend/distributed/planner/multi_physical_planner.c

```diff
@@ -2691,6 +2697,9 @@ SqlTaskList(Job *job)
 	List *fragmentCombinationList = FragmentCombinationList(rangeTableFragmentsList,
 															jobQuery, dependentJobList);
 
+	/* Adjust RelabelType and CoerceViaIO nodes that are improper for deparsing */
+	jobQuery = (Query *) AdjustImproperForDeparseNodes((Node *) jobQuery);
```

Comment:
note that query_tree_mutator makes a deep copy of the query tree, so the overhead of always doing this can be significant for complex queries

I assume that's why RelabelTypeToCollateExpr is so far used in the deparser

Reply:
I have added `QTW_DONT_COPY_QUERY` flag to the `query_tree_mutator` function such that it's modified in place

```C
return (Node *) query_tree_mutator((Query *) inputNode, AdjustImproperForDeparseNodes,
				   NULL, QTW_DONT_COPY_QUERY);
```

From Postgres:
```C
/*
 * query_tree_mutator --- initiate modification of a Query's expressions
 *
...
 * Normally the top-level Query node itself is copied, but some callers want
 * it to be modified in-place; they must pass QTW_DONT_COPY_QUERY in flags.
 * All modified substructure is safely copied in any case.
 */
Query *
query_tree_mutator(Query *query,
	           Node *(*mutator) (),
		   void *context,
		   int flags)
```

Reply:
Oops I missed `All modified substructure is safely copied in any case.`

Reply:
Well, that only takes care of the top-level Query struct and all the expression nodes below it are still copied.

Given the complexity of updating ruleutils, making the fix separately kind of makes sense, but it would be good to add a check (a walker that checks whether any relabel types exist).